crate/server/src/routes/google_cse/operations.rs

```diff
@@ -448,73 +471,527 @@ pub async fn private_key_decrypt(
+    let database_params = kms.get_sqlite_enc_secrets(&req_http)?;
+
+    let dek_data = cse_wrapped_key_decrypt(
+        request.wrapped_key,
+        UniqueIdentifier::TextString("google_cse".to_owned()),
+        user,
+        Some(resource_name.clone().into_bytes()),
+        kms,
+        database_params,
+        EncodingOption::NoEncoding,
+    )
+    .await?;
+
+    let base64_digest = compute_resource_key_hash(&resource_name, &perimeter_id, &dek_data)?;
+
+    Ok(DigestResponse {
+        resource_key_hash: base64_digest,
+    })
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PrivilegedWrapRequest {
+    pub authentication: String,
+    pub key: String,
+    pub perimeter_id: String,
+    pub resource_name: String,
+    pub reason: String,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct PrivilegedWrapResponse {
+    pub wrapped_key: String,
+}
+
+/// Returns a wrapped Data Encryption Key (DEK) and associated data. Use this method to encrypt data imported to Google Drive in bulk by a domain administrator.
+///
+/// See [doc](https://developers.google.com/workspace/cse/reference/privileged-wrap) and
+/// for more details, see [Encrypt & decrypt data](https://developers.google.com/workspace/cse/guides/encrypt-and-decrypt-data)
+///
+/// # Errors
+/// Will return `KmsError` if if authentication with tokens is incorrect, or if encryption fails
+pub async fn privileged_wrap(
+    request: PrivilegedWrapRequest,
+    cse_config: &Arc<Option<GoogleCseConfig>>,
+    kms: &Arc<KMSServer>,
+) -> KResult<PrivilegedWrapResponse> {
+    debug!("privileged-wrap: validate authentication token");
+    let user = validate_cse_authentication_token(&request.authentication, cse_config, true).await?;
+
+    debug!("privileged-wrap: wrap dek");
+    let resource_name = request.resource_name.into_bytes();
+    let encryption_request = Encrypt {
+        unique_identifier: Some(UniqueIdentifier::TextString("google_cse".to_owned())),
+        cryptographic_parameters: None,
+        data: Some(general_purpose::STANDARD.decode(&request.key)?.into()),
+        iv_counter_nonce: None,
+        correlation_value: None,
+        init_indicator: None,
+        final_indicator: None,
+        authenticated_encryption_additional_data: Some(resource_name),
+    };
+    let dek = encrypt(kms, encryption_request, &user, None).await?;
+
+    // re-extract the bytes from the key
+    let data = dek.data.ok_or_else(|| {
+        KmsError::InvalidRequest("Invalid wrapped key - missing data.".to_owned())
+    })?;
+    let iv_counter_nonce = dek.iv_counter_nonce.ok_or_else(|| {
+        KmsError::InvalidRequest("Invalid wrapped key - missing nonce.".to_owned())
+    })?;
+    let authenticated_encryption_tag = dek.authenticated_encryption_tag.ok_or_else(|| {
+        KmsError::InvalidRequest("Invalid wrapped key - authenticated encryption tag.".to_owned())
+    })?;
+
+    let wrapped_dek = [iv_counter_nonce, data, authenticated_encryption_tag].concat();
```

Comment:
```suggestion

    let mut wrapped_dek = Vec::with_capacity(iv_counter_nonce.len() + data.len() + authenticated_encryption_tag.len());
    wrapped_dek.extend_from_slice(&iv_counter_nonce);
    wrapped_dek.extend_from_slice(&data);
    wrapped_dek.extend_from_slice(&authenticated_encryption_tag);
```

I believe `extend_from_slice` does less copies than `concat`.

Reply:
I updated all the occurences